kompute: add backend registry / device interfaces and Q4_K shader

slp opened the pull request:
Get in line with the other backends by supporting the newer backend/device registry interfaces. While there, also add mul_mat_q4_k shader.

- [x] I have read the [contributing guidelines](https://github.com/ggerganov/llama.cpp/blob/master/CONTRIBUTING.md)
- Self-reported review complexity:
  - [ ] Low
  - [x] Medium
  - [ ] High

slp committed:
kompute: add backend registry / device interfaces

Get in line with the other backends by supporting the newer
backend/device registry interfaces.

Signed-off-by: Sergio Lopez <slp@redhat.com>

slaren commented on ggml/src/ggml-kompute.cpp:
```diff
@@ -2006,7 +2033,7 @@ ggml_backend_t ggml_backend_kompute_init(int device) {
         /* .guid      = */ ggml_backend_kompute_guid(),
         /* .interface = */ kompute_backend_i,
-        /* .device    = */ nullptr,
+        /* .device    = */ ggml_backend_reg_dev_get(ggml_backend_kompute_reg(), 0),
```
It's important that the device is set to the correct one

slp replied:
Fixed, thanks.

slaren commented on ggml/src/ggml-kompute.cpp:
```diff
@@ -1921,7 +1958,7 @@ ggml_backend_buffer_type_t ggml_backend_kompute_buffer_type(int device) {
             vec.push_back({
                 /* .iface   = */ ggml_backend_kompute_buffer_type_interface,
-                /* .device  = */ nullptr,
+                /* .device  = */ ggml_backend_reg_dev_get(ggml_backend_kompute_reg(), 0),
```
Same issue here.

slaren reviewed (commented):
I cannot review the other changes, I would suggest keeping them in a separate PR.

slp commented:
> I cannot review the other changes, I would suggest keeping them in a separate PR.

Thanks for the review. I've fixed the issues, cleaned up the code a little bit more, and dropped the commit adding the additional shader to deal with it in another PR.

slaren reviewed (approved):
The implementation of the backend reg and device interfaces looks good, can be merged after resolving the conflicts from #10026.

slp commented:
> The implementation of the backend reg and device interfaces looks good, can be merged after resolving the conflicts from #10026.

Thanks, I've just rebased it on top of current master.
